selftest.py

```diff
@@ -0,0 +1,138 @@
+    """The result of running a single self-test.
+
+    HasSelfTest.run_self_tests() returns a list of these
+    """
+
+    def __init__(self, name):
+        # Name of the test.
+        self.name = name
+
+        # Set to True when the test runs without raising an exception.
+        self.success = False
+
+        # The exception raised, if any.
+        self.exception = None
+
+        # The return value of the test method, assuming it ran to
+        # completion.
+        self.result = None
+
+        # Start time of the test.
+        self.start = datetime.datetime.utcnow()
+
+        # End time of the test.
+        self.end = None
+
+    def __repr__(self):
+        if self.exception:
+            if (isinstance(self.exception, IntegrationException)
+                and self.exception.debug_message):
+                exception = " exception=%r debug=%r" % (
+                    self.exception.message, self.exception.debug_message
+                )
+            else:
+                exception = " exception=%r" % self.exception
+        else:
+            exception = ""
+        return "<SelfTestResult: name=%r timing=%.2fsec success=%r%s result=%r>" % (
+            self.name, (self.end-self.start).total_seconds(), self.success,
+            exception, self.result
+        )
+
+
+class HasSelfTests(object):
+    """An object capable of verifying its own setup by running a
+    series of self-tests.
+    """
+
+    @classmethod
+    def run_self_tests(cls, _db, constructor_method=None, *args, **kwargs):
+        """Instantiate this class and call _run_self_tests on it.
+
+        :param _db: A database connection. Will be passed into
+        _run_self_tests. This connection may need to be used again
+        in *args, if the constructor needs it.
+
+        :param constructor_method: Method to use to instantiate the
+        class, if different from the default constructor.
+
+        :param args: Positional arguments to pass into the constructor.
+        :param kwargs: Keyword arguments to pass into the constructor.
+
+        :return: An iterator of SelfTestResult objects, starting with
+        the attempt to instantiate the test class in the first place.
+        """
+        constructor_method = constructor_method or cls
+        result = SelfTestResult("Initial setup.")
+        instance = None
+        try:
+            instance = constructor_method(*args, **kwargs)
+            result.success = True
+            result.result = instance
+        except Exception, e:
+            result.exception = e
+            result.success = False
+        finally:
+            result.end = datetime.datetime.utcnow()
+        yield result
+
+        if instance:
+            for result in instance._run_self_tests(_db):
+                yield result
+
+    def _run_self_tests(self, _db):
+        """Run a series of self-tests.
+
+        :return: A list of SelfTestResult objects.
+        """
+        raise NotImplementedError()
```

Comment:
I'm wondering if this should return a self test failure object that says "self tests have not been implemented for this integration". Otherwise, we'd likely get a 500 when trying to run tests from the admin interface, so we'd need some way to tell the admin interface in advance which integrations have self tests.

Reply:
I guess we need that anyway, since integrations that don't have them yet won't extend this class.

Reply:
I was thinking the admin interface wouldn't offer the self-test functionality for any class that didn't implement HasSelfTests. It needs to know how to instantiate the class, so there's no way to write fully general code anyway.